### PR TITLE
in case student's files shadow packages

### PR DESCRIPTION
**Pull request by dmalan**

### CODE REVIEW

**Comment by dmalan** [comment]
@crossroads1112 @brianyu28 @kzidane per email discussion!

**Comment by crossroads1112** [comment]
Sorry about that! Just to be safe, we may want to do the importing in a "try" and restore the path in a "finally", just to make sure that we always restore the students' path even if an exception is thrown (this would only matter if the student tried to handle the exception, so it's unlikely that this would ever come up, but it seems like good practice to make sure we always restore `sys.path`).

**Comment by dmalan** [comment]
@crossroads1112 hm, do we really want to swallow in `__init__.py` any exceptions that our own code accidentally raises? atm, if any such exception is raised, that should indicate a bug in the library, which we're best off seeing in the wild?

**Comment by crossroads1112** [comment]
I don't mean we'd catch the exception, just that we'd clean up if there was one; the exception is still raised even when there is a finally block e.g.

```
path = sys.path
sys.path = ...
try:
    import foo
    import bar
    # ...
finally:
    sys.path = path
```
if an exception is raised, the cleanup code gets run, but the exception still propagates.

**Comment by dmalan** [comment]
TIL. How's it look now?

File: src/cs50/__init__.py [hunk 1]
```diff
@@ -1,34 +1,48 @@
...
-    http://dangerontheranger.blogspot.com/2012/07/how-to-use-sysmetapath-with-python.html
-    """
+    # Replace flask's logger
+    from . import flask
 
-    def find_module(self, fullname, path=None):
-        if fullname == "cs50.SQL":
-            return self
-        return None
+    class CustomImporter(object):
```
Comment (crossroads1112) [inline src/cs50/__init__.py]: Aesthetically we could maybe move this above the try block (so it isn't indented an extra level), but not a big deal

File: src/cs50/__init__.py [hunk 2]
```diff
@@ -1,34 +1,48 @@
...
+    sys.path = [p for p in sys.path if p not in ("", os.getcwd())]
 
-class CustomImporter(object):
-    """
-    Import cs50.SQL lazily so that rest of library can be used without SQLAlchemy installed.
+    # Import cs50_*
+    from .cs50 import eprint, get_char, get_float, get_int, get_string
+    try:
+        from .cs50 import get_long
+    except:
```
Comment (crossroads1112) [inline src/cs50/__init__.py]: Should probably be `except Exception:` unless we want to catch `SystemExit` et al.

File: src/cs50/__init__.py [hunk 3]
```diff
@@ -1,34 +1,48 @@
...
-    from .cs50 import get_long
-except:
-    pass
 
-from . import flask
+    # Save student's sys.path
+    path = sys.path[:]
 
+    # In case student has files that shadow packages
+    sys.path = [p for p in sys.path if p not in ("", os.getcwd())]
```
Comment (crossroads1112) [inline src/cs50/__init__.py]: Per the docs, the directory in which the script is run (or `""` if in the interpreter) is always `sys.path[0]`. With this in mind, I think we could just do `del sys.path[0]` for the same effect (and thus avoid having to reconstruct the whole list). 

Alternatively we could do e.g.
```
try:
    cwd = sys.path.pop(0)
   # ...
finally:
    sys.path.insert(0, cwd)
```
which avoids having to store two copies in the first place.

Comment (dmalan) [reply src/cs50/__init__.py]: Hm, technically, a student could have prepended a path of their own? Feels more robust not to make any assumptions? And cost of reconstruction is pretty trivial.

File: src/cs50/__init__.py (hunk 2 above)
Comment (dmalan) [reply src/cs50/__init__.py]: TIL

File: src/cs50/__init__.py (hunk 1 above)
Comment (dmalan) [reply src/cs50/__init__.py]: Not opposed.

**Comment by dmalan** [comment]
:shipit:, @crossroads1112? https://github.com/cs50/python-cs50/pull/46/commits/71b92c60c42962fd2488835898467627b2925026

**Pull request merged**